kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java

```diff
@@ -203,11 +203,34 @@ public CloseableIterator<FilteredColumnarBatch> getAddFilesAsColumnarBatches(
   protected Tuple2<Protocol, Metadata> loadTableProtocolAndMetadata(
       Engine engine, Optional<SnapshotHint> snapshotHint, long snapshotVersion) {
 
-    // Exit early if the hint already has the info we need
+    // If the snapshot hint is provided and matches the expected version, load the P&M from version hint.
+    // If not, e.g. for the case of time travel, skip use version hint and fallback to CRC file or log replay.
     if (snapshotHint.isPresent() && snapshotHint.get().getVersion() == snapshotVersion) {
       return new Tuple2<>(snapshotHint.get().getProtocol(), snapshotHint.get().getMetadata());
     }
 
+    // Compute the lower bound for the CRC search
+    // If the snapshot hint is present, we can use it as the lower bound for the CRC search.
+    // TODO: this can be further improved to make the lower bound until the checkpoint version
+    Optional<Long> crcSearchLowerBound = snapshotHint.map(SnapshotHint::getVersion);
+    Optional<CRCInfo> crcInfoOpt =
+        ChecksumReader.getCRCInfo(engine, logSegment.logPath, snapshotVersion, crcSearchLowerBound);
+    if (crcInfoOpt.isPresent()) {
+      CRCInfo crcInfo = crcInfoOpt.get();
+      if (crcInfo.getVersion() == snapshotVersion) {
+        // CRC is related to the desired snapshot version. Load protocol and metadata from CRC.
+        return new Tuple2<>(crcInfo.getProtocol(), crcInfo.getMetadata());
+      }
+      // We found the protocol and metadata in a version older than the one we are looking
+      // for. We need to replay the actions to get the latest protocol and metadata, but
+      // update the hint to read the actions from the version we found to check if the
+      // protocol and metadata are updated in the versions after the one we found.
+      // Building a new snapshotHit for determining the time to end the log replay loop.
+      snapshotHint =
+          Optional.of(
+              new SnapshotHint(crcInfo.getVersion(), crcInfo.getProtocol(), crcInfo.getMetadata()));
```

Comment:
nit: maybe `SnapshotHint.fromCRCInfo`?

Reply:
I'm okay with how it is for now. Im starting to see lots of coupling between SnapshotHint and CRCInfo. Let's wait to introduce static helpers like that, and perhaps instead re-evaluate the relationship between these two classes (in a later PR)